src/mainloop.rkt

```diff
@@ -396,15 +396,37 @@
   (extract!))
 
 (define (pareto-regimes sorted ctx)
-  (let loop ([alts sorted])
+  (define branch-exprs
+    (if (flag-set? 'reduce 'branch-expressions)
+        (exprs-to-branch-on sorted ctx)
+        (context-vars ctx)))
+  (define err-lsts (batch-errors (map alt-expr sorted) (*pcontext*) ctx))
+  (define init-errs (for/hash ([bexpr branch-exprs]) (values bexpr -1)))
+  (let loop ([alts sorted] [errs init-errs] [try-first (first branch-exprs)])
     (cond
      [(null? alts) '()]
      [(= (length alts) 1) (list (car alts))]
      [else
-      (define opt (infer-splitpoints alts ctx))
+      (define recomputed-branch-exprs
+        (if (flag-set? 'reduce 'branch-expressions)
+            (exprs-to-branch-on alts ctx)
+            (context-vars ctx)))
+      
+      (define recomp-errs 
+        (for/hash ([bexpr recomputed-branch-exprs]) (values bexpr (hash-ref errs bexpr))))
+
+      ;; if after recomputing branch-expressions, the previously chosen branch-expression no longer exists, use a
+      ;; different one.
+      (define recomp-try 
+        (if (member try-first recomputed-branch-exprs) 
+            try-first 
+            (argmin (curry hash-ref recomp-errs) recomputed-branch-exprs)))
```

Comment:
I think if we do the sort-branches-by-lowest-cerr thing I suggested above, then we don't need to track the previous best branch (it'll just naturally sort to the front because it has the lowest cerr) so then we won't need this code either.